keps/sig-network/20191220-apiserver-url-envvar.md

```diff
@@ -0,0 +1,294 @@
+
+### Upgrade / Downgrade Strategy
+
+If applicable, how will the component be upgraded and downgraded? Make sure this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing cluster required to make on upgrade in order to keep previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing cluster required to make on upgrade in order to make use of the enhancement?
+
+### Version Skew Strategy
+
+If applicable, how will the component handle version skew with other components? What are the guarantees? Make sure
+this is in the test plan.
+
+Consider the following in developing a version skew strategy for this enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and in the kubelet? How does an n-2 kubelet without this feature available behave when this feature is used?
+- Will any other components on the node change? For example, changes to CSI, CRI or CNI may require updating that component before the kubelet.
+
+## Implementation History
+
+Major milestones in the life cycle of a KEP should be tracked in `Implementation History`.
+Major milestones might include
+
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+
+## Drawbacks
+
+The biggest drawbacks of this proposal are that several
+[alternatives](#alternatives) are also appealing.  It is not clear
+that _this_ proposal is the right tradeoff since it is neither the
+most minimal, nor the "best".
+
+## Alternatives
+
+There are several fine alternatives to this proposal:
+- Continue to use KUBERNETES_SERVICE_HOST, but change it to
+  `kubernetes.default` or other hostname instead of an IP (v4 or v6)
+  address.
+
+  Cons: This requires DNS, but is an even simpler change than this KEP.
+
+- Declare that the endpoint is always `kubernetes.default:443` (ie:
+  ignore environment variables)
+
+  Cons: Declaring host is probably ok. Declaring port is probably not.
+
+- Declare the endpoint is always `kubernetes.kube-system:443`
+  (or some similar new/better clusterdns hostname)
+
+- Declare the endpoint is always `_kubernetes._tcp` DNS `SRV`.
+  Cool.  Unfortunately SRV is not trivial to use in a URL.
+
+- Declare a well-known (anycast) IPv4/IPv6 address and port.
+  Like some cloud providers did for 169.254.169.254 instance metadata.
+
+  Cons: This has numerous issues for instances. We should not repeat
+  these mistakes given the alternatives available.
+
+- Write a file to a well-known path in the pod containing the
```

Comment:
A file has the benefit that it can be changed at runtime. Environment variables require restart of the containers to reload.

Reply:
To do this as a file we would need to use downward API into a volume.  Users would have to opt-in to it, which makes it less than ideal for being the backstop for clients to find the master

Reply:
Well, we _could_ mandate it as a new CRI requirement (hand wave) - and not use downward API.

To the original point, I think many clients will only read it once at startup so they can (eg) cache/reuse a persistent HTTPS connection.  So I'm guessing changing it at runtime won't work anyway .. unless we _force_ frequent changes on everyone by being deliberately perverse (cycle ClusterIP addresses just because, etc).

Reply:
Agree with all points.  I don't see much value in making this mutable -- if that becomes an issue, we can discuss downward API support for it.